crates/sui-graphql-rpc/src/types/transaction_block_effects.rs

```diff
@@ -106,40 +151,115 @@ impl TransactionBlockEffects {
     }
 
     /// Shared objects that are referenced by but not changed by this transaction.
-    async fn unchanged_shared_objects(&self) -> Option<Vec<UnchangedSharedObject>> {
-        Some(
-            self.native
-                .input_shared_objects()
-                .into_iter()
-                .filter_map(|input| UnchangedSharedObject::try_from(input).ok())
-                .collect(),
-        )
+    async fn unchanged_shared_objects(
+        &self,
+        ctx: &Context<'_>,
+        first: Option<u64>,
+        after: Option<CUnchangedSharedObject>,
+        last: Option<u64>,
+        before: Option<CUnchangedSharedObject>,
+    ) -> Result<Connection<String, UnchangedSharedObject>> {
+        let page = Page::from_params(ctx.data_unchecked(), first, after, last, before)?;
+        let mut connection = Connection::new(false, false);
+
+        // TODO: this iterates through its shared_objects field and collects.
+        // We can consider doing that collection here so we don't double dip.
```

Comment:
I think this is OK -- what we're trying to protect against here is more expensive sub-queries, than the work of finding the input shared objects -- fields on transaction block effects are already subject to limits.